### ISSUE TITLE
Compile error: src/freebsd/btop_collect.cpp:663:67: error: '_PATH_DEVnullptr' was not declared in this scope

### ISSUE THREAD

**Issue by tessus** [aristocratos/btop#593]
Updated a bsd dev machine to FreeBSD 13.2 and get the following error when compiling btop:

```
$ gmake

 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Makefile v1.6
PLATFORM   ?| FreeBSD
ARCH       ?| x86_64
CXX        ?| g++12 (12.2.0)
THREADS    :| 4
REQFLAGS   !| -std=c++20
WARNFLAGS  :| -Wall -Wextra -pedantic
OPTFLAGS   :| -O2 -ftree-vectorize -flto=4
LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -D_FILE_OFFSET_BITS=64 -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -Wl,--fatal-warnings -s -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc11
CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)
LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)

Building btop++ (v1.2.13) FreeBSD x86_64
Compiling src/freebsd/btop_collect.cpp
src/freebsd/btop_collect.cpp: In function 'Mem::mem_info& Mem::collect(bool)':
src/freebsd/btop_collect.cpp:663:67: error: '_PATH_DEVnullptr' was not declared in this scope
  663 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
      |                                                                   ^~~~~~~~~~~~~~~~
src/freebsd/btop_collect.cpp: In function 'std::vector<Proc::proc_info>& Proc::collect(bool)':
src/freebsd/btop_collect.cpp:1160:67: error: '_PATH_DEVnullptr' was not declared in this scope
 1160 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
      |                                                                   ^~~~~~~~~~~~~~~~
src/freebsd/btop_collect.cpp:1182:60: warning: the address of 'kinfo_proc::ki_comm' will never be NULL [-Waddress]
 1182 |                                         if (kproc->ki_comm == nullptr or kproc->ki_comm == "idle"s) {
      |                                             ~~~~~~~~~~~~~~~^~~~~~~~~~
In file included from src/freebsd/btop_collect.cpp:39:
/usr/include/sys/user.h:182:17: note: 'kinfo_proc::ki_comm' declared here
  182 |         char    ki_comm[COMMLEN+1];     /* command name */
      |                 ^~~~~~~
gmake: *** [Makefile:308: obj/freebsd/btop_collect.o] Error 1
```

I am currently trying to compile git master (8a33aab). I know that 4689938 compiles w/o issues.
I'll do a git bisect when I get to it later this evening.

**Comment by aristocratos** [comment]
@tessus 
Changing `-Wl,-rpath=/usr/local/lib/gcc11` to `-Wl,-rpath=/usr/local/lib/gcc12` in the Makefile should fix it.

Are gonna change it so the path updates to currently used compiler.

**Comment by aristocratos** [comment]
@tessus 
I apparently didn't look carefully enough :P

It seems `_PATH_DEVNULL` was changed by mistake when replacing `NULL` with `nullptr`.
So that should be the issue.

But does it compile fine for you otherwise with gcc12?

**Comment by tessus** [comment]
yes, sort of. Here is the result

```
e4abcefbf92e5d94ad169e1e47c0fbec7279fa6f is the first bad commit
commit e4abcefbf92e5d94ad169e1e47c0fbec7279fa6f
Author: nobounce <steffen.winter@proton.me>
Date:   Wed Jul 26 16:19:17 2023 +0200

    Use nullptr instead of NULL.

    See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2431.pdf
    TLDR: NULL is of type int and relies on proper implicit pointer
    conversion which may lead to issues when using overloaded functions

    It is also considered a 'best practise' for modern C++ and
    conveys the programmers intention more precisly.

 src/btop.cpp                 | 26 +++++++--------
 src/btop_draw.cpp            |  2 +-
 src/btop_input.hpp           |  2 +-
 src/btop_tools.cpp           | 12 +++----
 src/freebsd/btop_collect.cpp | 76 ++++++++++++++++++++++----------------------
 src/linux/btop_collect.cpp   | 10 +++---
 src/osx/btop_collect.cpp     | 42 ++++++++++++------------
 7 files changed, 85 insertions(+), 85 deletions(-)
```

But I can see that there is a warning when compiling with g++12:

```
Compiling src/freebsd/btop_collect.cpp
src/freebsd/btop_collect.cpp: In function 'std::vector<Proc::proc_info>& Proc::collect(bool)':
src/freebsd/btop_collect.cpp:1182:60: warning: the address of 'kinfo_proc::ki_comm' will never be NULL [-Waddress]
 1182 |                                         if (kproc->ki_comm == NULL or kproc->ki_comm == "idle"s) {
      |                                                            ^
In file included from src/freebsd/btop_collect.cpp:39:
/usr/include/sys/user.h:182:17: note: 'kinfo_proc::ki_comm' declared here
  182 |         char    ki_comm[COMMLEN+1];     /* command name */
      |                 ^~~~~~~
```

**Comment by tessus** [comment]
I am not very good with FreeBSD. I have it only installed in a VM to compile btop for my OPNsense box.

**Comment by tessus** [comment]
> Changing -Wl,-rpath=/usr/local/lib/gcc11 to -Wl,-rpath=/usr/local/lib/gcc12 in the Makefile should fix it.

Nope, it did not help. Same error:

```
gmake

 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Makefile v1.6
PLATFORM   ?| FreeBSD
ARCH       ?| x86_64
CXX        ?| g++12 (12.2.0)
THREADS    :| 4
REQFLAGS   !| -std=c++20
WARNFLAGS  :| -Wall -Wextra -pedantic
OPTFLAGS   :| -O2 -ftree-vectorize -flto=4
LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -D_FILE_OFFSET_BITS=64 -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -static -Wl,--fatal-warnings -s -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc12
CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)
LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)

Building btop++ (v1.2.13) FreeBSD x86_64
Compiling src/freebsd/btop_collect.cpp
src/freebsd/btop_collect.cpp: In function 'Mem::mem_info& Mem::collect(bool)':
src/freebsd/btop_collect.cpp:663:67: error: '_PATH_DEVnullptr' was not declared in this scope
  663 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
      |                                                                   ^~~~~~~~~~~~~~~~
src/freebsd/btop_collect.cpp: In function 'std::vector<Proc::proc_info>& Proc::collect(bool)':
src/freebsd/btop_collect.cpp:1160:67: error: '_PATH_DEVnullptr' was not declared in this scope
 1160 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
      |                                                                   ^~~~~~~~~~~~~~~~
src/freebsd/btop_collect.cpp:1182:60: warning: the address of 'kinfo_proc::ki_comm' will never be NULL [-Waddress]
 1182 |                                         if (kproc->ki_comm == nullptr or kproc->ki_comm == "idle"s) {
      |                                             ~~~~~~~~~~~~~~~^~~~~~~~~~
In file included from src/freebsd/btop_collect.cpp:39:
/usr/include/sys/user.h:182:17: note: 'kinfo_proc::ki_comm' declared here
  182 |         char    ki_comm[COMMLEN+1];     /* command name */
      |                 ^~~~~~~
gmake: *** [Makefile:308: obj/freebsd/btop_collect.o] Error 1
```

**Comment by tessus** [comment]
> It seems _PATH_DEVNULL was changed by mistake when replacing NULL with nullptr.

Missed that part. Let me quickly change that back.

**Comment by aristocratos** [comment]
@tessus 
You can try with the latest commit I just pushed.
Should fix both the nullptr changes and the Makefile.

**Comment by tessus** [comment]
Hmm, it compiles, but now fails when linking:

```
Linking and optimizing binary...
/usr/local/bin/ld: /usr/lib/libkvm.a(kvm.o): in function `_kvm_read_kernel_ehdr':
/usr/src/lib/libkvm/kvm.c:85: undefined reference to `elf_version'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:89: undefined reference to `elf_begin'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:94: undefined reference to `elf_kind'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:98: undefined reference to `gelf_getehdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:103: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:91: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:99: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:100: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/lib/libkvm.a(kvm_private.o): in function `_kvm_read_core_phdrs':
/usr/src/lib/libkvm/kvm_private.c:169: undefined reference to `elf_begin'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:174: undefined reference to `elf_kind'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:178: undefined reference to `gelf_getclass'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:182: undefined reference to `gelf_getehdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:195: undefined reference to `elf_getphdrnum'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:207: undefined reference to `gelf_getphdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:213: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:219: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:171: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:(.text+0x509): undefined reference to `elf_errmsg'
collect2: error: ld returned 1 exit status
gmake: *** [Makefile:297: btop] Error 1
```

**Comment by aristocratos** [comment]
@tessus 
Did you run `gmake distclean` before compiling again?

What does the complete output look like?

**Comment by tessus** [comment]
I ran `gmake clean`, but now I ran distclean as well. Here is the full output:

```
$ gmake distclean
Removing: built objects...
Removing: built binaries...
$ gmake

 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Makefile v1.6
PLATFORM   ?| FreeBSD
ARCH       ?| x86_64
CXX        ?| g++12 (12.2.0)
THREADS    :| 4
REQFLAGS   !| -std=c++20
WARNFLAGS  :| -Wall -Wextra -pedantic
OPTFLAGS   :| -O2 -ftree-vectorize -flto=4
LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -D_FILE_OFFSET_BITS=64 -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -static -Wl,--fatal-warnings -s -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc12
CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)
LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)

Building btop++ (v1.2.13) FreeBSD x86_64
Compiling src/btop.cpp
Compiling src/btop_draw.cpp
Compiling src/btop_config.cpp
Compiling src/btop_input.cpp
10%  -> obj/btop_input.o              (512BiB) (04s)
20%  -> obj/btop_config.o             (512BiB) (04s)
Compiling src/btop_menu.cpp
Compiling src/btop_shared.cpp
40%  -> obj/btop.o                    (512BiB) (05s)
40%  -> obj/btop_draw.o               (512BiB) (05s)
Compiling src/btop_theme.cpp
Compiling src/btop_tools.cpp
50%  -> obj/btop_shared.o             (512BiB) (05s)
Compiling src/freebsd/btop_collect.cpp
70%  -> obj/btop_menu.o               (512BiB) (05s)
70%  -> obj/btop_theme.o              (512BiB) (03s)
80%  -> obj/btop_tools.o              (512BiB) (03s)
90%  -> obj/freebsd/btop_collect.o    (512BiB) (05s)

Linking and optimizing binary...
/usr/local/bin/ld: /usr/lib/libkvm.a(kvm.o): in function `_kvm_read_kernel_ehdr':
/usr/src/lib/libkvm/kvm.c:85: undefined reference to `elf_version'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:89: undefined reference to `elf_begin'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:94: undefined reference to `elf_kind'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:98: undefined reference to `gelf_getehdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:103: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:91: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:99: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm.c:100: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/lib/libkvm.a(kvm_private.o): in function `_kvm_read_core_phdrs':
/usr/src/lib/libkvm/kvm_private.c:169: undefined reference to `elf_begin'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:174: undefined reference to `elf_kind'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:178: undefined reference to `gelf_getclass'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:182: undefined reference to `gelf_getehdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:195: undefined reference to `elf_getphdrnum'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:207: undefined reference to `gelf_getphdr'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:213: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:219: undefined reference to `elf_end'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:171: undefined reference to `elf_errmsg'
/usr/local/bin/ld: /usr/src/lib/libkvm/kvm_private.c:(.text+0x509): undefined reference to `elf_errmsg'
collect2: error: ld returned 1 exit status
gmake: *** [Makefile:297: btop] Error 1
```

**Comment by aristocratos** [comment]
@tessus 
Static build won't work on FreeBSD.

**Comment by tessus** [comment]
Since when and why? It compiles perfectly fine with https://github.com/aristocratos/btop/commit/468993817ad62155631442307f5bdb945a7214f9 and the Makefile fix to add `-static`

**Comment by tessus** [comment]
This is strange. When I use commit hash https://github.com/aristocratos/btop/commit/468993817ad62155631442307f5bdb945a7214f9 and my previously modified version of the [Makefile](https://evermeet.cx/paste/Makefile.OMeQ.txt) it links perfectly fine.

**Comment by aristocratos** [comment]
I've never gotten static linking working on FreeBSD (with btop), it's the main reason I don't provide any release binaries for FreeBSD.

**Comment by tessus** [comment]
Ok, so now I compiled and linked git master statically on FreeBSD with this version of the Makefile: https://evermeet.cx/paste/Makefile.OMeQ.txt

```
$ gmake

 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Makefile v1.5
PLATFORM   ?| FreeBSD
ARCH       ?| x86_64
CXX        ?| g++11 (11.3.0)
THREADS    :| 4
REQFLAGS   !| -std=c++20
WARNFLAGS  :| -Wall -Wextra -pedantic
OPTFLAGS   :| -O2 -ftree-vectorize -flto=4
LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -Wl,--fatal-warnings -s -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc11
CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)
LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)

Building btop++ (v1.2.13) FreeBSD x86_64
Compiling src/btop_draw.cpp
Compiling src/btop_config.cpp
Compiling src/btop.cpp
Compiling src/btop_input.cpp
10%  -> obj/btop_input.o              (512BiB) (03s)
20%  -> obj/btop_config.o             (512BiB) (03s)
Compiling src/btop_menu.cpp
Compiling src/btop_shared.cpp
30%  -> obj/btop_draw.o               (512BiB) (04s)
40%  -> obj/btop.o                    (512BiB) (05s)
Compiling src/btop_theme.cpp
Compiling src/btop_tools.cpp
50%  -> obj/btop_shared.o             (512BiB) (04s)
60%  -> obj/btop_menu.o               (512BiB) (04s)
Compiling src/freebsd/btop_collect.cpp
70%  -> obj/btop_theme.o              (512BiB) (03s)
80%  -> obj/btop_tools.o              (512BiB) (04s)
90%  -> obj/freebsd/btop_collect.o    (512BiB) (06s)

Linking and optimizing binary...
100% -> bin/btop                      (512BiB) (07s)

Build complete in (22s)
```

However, this Makefile seems to use g++11.

**Comment by aristocratos** [comment]
@tessus 
I get the same error (the same you posted above) regardless which compiler I use. Tried gcc10, 11 and 12.
I'm also on FreeBSD 13.2.

So I'm really surprised it compiles for you on gcc11.

**Comment by tessus** [comment]
Did you try the Makefile I uploaded? It doesn't work for me with the Mskefile in the source tree either.

**Comment by aristocratos** [comment]
You can just use `gmake CXX=g++11 STATIC=true` which should get you the same result.

**Comment by tessus** [comment]
A while back (I can't recall the commit hash) I had to modify the Makefile to add `-static`. Either way, the Makefile I posted works. The Makefile in the source tree with `gmake CXX=g++11 STATIC=true` does not work.

**Comment by aristocratos** [comment]
@tessus 
Looking at your Makefile, you removed the `-static` flag, thus not doing a complete static build, but just building libgcc and libstdc++ statically.

Try `gmake ADDFLAGS="-static-libgcc -static-libstdc++ -DSTATIC_BUILD"`, works fine with gcc12 and adds same flags as your modified Makefile. But it isn't a fully static build, and can only be guaranteed to run on FreeBSD 13.2.

**Comment by tessus** [comment]
Ah, that rings a bell. yep, I think you are correct. It was not completely static, but I am ok with a not completely static build. For me it was only important that I didn't have to install anything but the binary on my OPNsense box. I keep my FreebSD VM at the same level as my OPNsense, so all is good for me. But the command you posted did not work for me:

For some reason the `ADDFLAGS="-static-libgcc -static-libstdc++ -DSTATIC_BUILD"` does not work. It compiles, but they are not linked statically:

```
$ gmake ADDFLAGS="-static-libgcc -static-libstdc++ -DSTATIC_BUILD"

 ██████╗ ████████╗ ██████╗ ██████╗
 ██╔══██╗╚══██╔══╝██╔═══██╗██╔══██╗   ██╗    ██╗
 ██████╔╝   ██║   ██║   ██║██████╔╝ ██████╗██████╗
 ██╔══██╗   ██║   ██║   ██║██╔═══╝  ╚═██╔═╝╚═██╔═╝
 ██████╔╝   ██║   ╚██████╔╝██║        ╚═╝    ╚═╝
 ╚═════╝    ╚═╝    ╚═════╝ ╚═╝      Makefile v1.6
PLATFORM   ?| FreeBSD
ARCH       ?| x86_64
CXX        ?| g++12 (12.2.0)
THREADS    :| 4
REQFLAGS   !| -std=c++20
WARNFLAGS  :| -Wall -Wextra -pedantic
OPTFLAGS   :| -O2 -ftree-vectorize -flto=4
LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -D_FILE_OFFSET_BITS=64 -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -s -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc12
CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)
LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS)

Building btop++ (v1.2.13) FreeBSD x86_64

Linking and optimizing binary...
100% -> bin/btop                      (512BiB) (08s)

Build complete in (08s)
[tessus@bsd-dev ~/btop]$ ldd bin/btop
bin/btop:
	libstdc++.so.6 => /usr/local/lib/gcc12/libstdc++.so.6 (0x780da400000)
	libkvm.so.7 => /lib/libkvm.so.7 (0x780dbbf4000)
	libdevstat.so.7 => /lib/libdevstat.so.7 (0x780db09c000)
	libm.so.5 => /lib/libm.so.5 (0x780dc463000)
	libthr.so.3 => /lib/libthr.so.3 (0x780dd43c000)
	libc.so.7 => /lib/libc.so.7 (0x780dec6a000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x780dd4a9000)
	libelf.so.2 => /lib/libelf.so.2 (0x780de23b000)
	[vdso] (0x7fffffffe5d0)
```

This is why I modified the Makefile and exported STATIC=true. When I use my Makefile AND use STATIC=true the resulting file looks like this:

```
$ ldd bin/btop
bin/btop:
	libkvm.so.7 => /lib/libkvm.so.7 (0x2d6b42c6c000)
	libdevstat.so.7 => /lib/libdevstat.so.7 (0x2d6b41b91000)
	libm.so.5 => /lib/libm.so.5 (0x2d6b425dd000)
	libthr.so.3 => /lib/libthr.so.3 (0x2d6b4360f000)
	libc.so.7 => /lib/libc.so.7 (0x2d6b44283000)
	libelf.so.2 => /lib/libelf.so.2 (0x2d6b45c1e000)
	[vdso] (0x7fffffffe5d0)
```

**Comment by aristocratos** [comment]
@tessus 
The "STATIC=true" is just a flag I added to the Makefile and the only thing it does is adds:
`-static-libgcc -static-libstdc++ -DSTATIC_BUILD -static -Wl,--fatal-warnings` to the compiler.
Nothing else.

Look at the LDCXXFLAGS in the output.
With `gmake ADDFLAGS="-static-libgcc -static-libstdc++ -DSTATIC_BUILD"` they're just added manually instead and without the `-static` flag. The only difference with the latest Makefile is that  `-D_FILE_OFFSET_BITS=64` was added which makes no difference for static linkage.

The reason it's not corretly linked is beacuse `-lstdc++` is present also in the flags.

To completely set the flags manually try:
`gmake LDCXXFLAGS="-pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -s -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc12"`

```bash
 ldd ./bin/btop
./bin/btop:
        libkvm.so.7 => /lib/libkvm.so.7 (0x85acd6e2000)
        libdevstat.so.7 => /lib/libdevstat.so.7 (0x85ace4c1000)
        libm.so.5 => /lib/libm.so.5 (0x85aceeef000)
        libthr.so.3 => /lib/libthr.so.3 (0x85ad0a24000)
        libc.so.7 => /lib/libc.so.7 (0x85acef83000)
        libelf.so.2 => /lib/libelf.so.2 (0x85acfc60000)
        [vdso] (0x7ffffffff000)
```

**Comment by tessus** [comment]
Yes, awesome. `gmake LDCXXFLAGS="-pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -static-libgcc -static-libstdc++ -DSTATIC_BUILD -s -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc12"` did the trick.

I now have a binary compiled with g++12 and the Makefile from the source tree:

```
$ ldd bin/btop
bin/btop:
	libkvm.so.7 => /lib/libkvm.so.7 (0x912bc9af000)
	libdevstat.so.7 => /lib/libdevstat.so.7 (0x912bb8dd000)
	libm.so.5 => /lib/libm.so.5 (0x912bbd59000)
	libthr.so.3 => /lib/libthr.so.3 (0x912bd5c1000)
	libc.so.7 => /lib/libc.so.7 (0x912bdf58000)
	libelf.so.2 => /lib/libelf.so.2 (0x912be3a6000)
	[vdso] (0x7fffffffe5d0)
```

**Comment by tessus** [comment]
Thanks for all the help. I think you can close this issue now.

**Comment by aristocratos** [comment]
@tessus 
Np :)

The Makefile is updated in the latest commit, so now you should only need to use `STATIC=true` for a "partial" static build on FreeBSD.

**Comment by tessus** [comment]
Yes, I can confirm that this is the case. Thanks a bunch.

**Comment by nobounce** [comment]
I'm wondering why this wasn't catched in CI. The PR and the actual merge both failed with
```
 PLATFORM   ?| FreeBSD
  ARCH       ?| x86_64
  CXX        ?| g++11 (11.3.0)
  THREADS    :| 3
  REQFLAGS   !| -std=c++20
  WARNFLAGS  :| -Wall -Wextra -pedantic
  OPTFLAGS   :| -O2 -ftree-vectorize -flto=3
  LDCXXFLAGS :| -pthread -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -D_FILE_OFFSET_BITS=64 -fexceptions -fstack-clash-protection -fcf-protection -fstack-protector -lstdc++ -lm -lkvm -ldevstat -Wl,-rpath=/usr/local/lib/gcc11
  CXXFLAGS   +| $(REQFLAGS) $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS) 
  LDFLAGS    +| $(LDCXXFLAGS) $(OPTFLAGS) $(WARNFLAGS) 
  
  Building btop++ (v1.2.13) FreeBSD x86_64
  Compiling src/btop.cpp
  Compiling src/btop_config.cpp
  Compiling src/btop_draw.cpp
  10%-> obj/btop_draw.o (2.3MiB) (50s)
  Compiling src/btop_input.cpp
  20%-> obj/btop_input.o (1.4MiB) (11s)
  Compiling src/btop_menu.cpp
  30%-> obj/btop_menu.o (2.0MiB) (10s)
  Compiling src/btop_shared.cpp
  40%-> obj/btop_shared.o (1.7MiB) (13s)
  Compiling src/btop_theme.cpp
  50%-> obj/btop_theme.o (1.4MiB) (11s)
  Compiling src/btop_tools.cpp
  60%-> obj/btop_tools.o (1.4MiB) (13s)
  Compiling src/freebsd/btop_collect.cpp
  src/freebsd/btop_collect.cpp: In function 'Mem::mem_info& Mem::collect(bool)':
  src/freebsd/btop_collect.cpp:663:67: error: '_PATH_DEVnullptr' was not declared in this scope
    663 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
        |                                                                   ^~~~~~~~~~~~~~~~
  src/freebsd/btop_collect.cpp: In function 'std::vector<Proc::proc_info>& Proc::collect(bool)':
  src/freebsd/btop_collect.cpp:1160:67: error: '_PATH_DEVnullptr' was not declared in this scope
   1160 |                         Shared::kvm_openfiles_wrapper kd(nullptr, _PATH_DEVnullptr, nullptr, O_RDONLY, buf);
        |                                                                   ^~~~~~~~~~~~~~~~
  **gmake: *** [Makefile:308: obj/freebsd/btop_collect.o] Error 1**
  gmake: *** Waiting for unfinished jobs....
  70%-> obj/btop_config.o (1.4MiB) (02m:07s)
  80%-> obj/btop.o (2.8MiB) (02m:19s)
```

but both were still flagged OK.
